src/encoding_transcode.rs

```diff
@@ -0,0 +1,432 @@
+        // BOM marker bytes and if it does, remove it for the final output.
+        //
+        // We can choose not to strip the BOM marker and keep it as is, but the
+        // presence of these extra bytes can throw off any downstream processing
+        // we do on the output, and rather than handling it specially on each
+        // processing, we handle it centrally here. Also, the BOM does not serve
+        // any more use for us, since the source encoding is already pre-identified
+        // as part of decoder initialization.
+        if output
+            .get(..BOM_UTF8_LEN)
+            .map_or(false, |start| start == BOM_UTF8)
+        {
+            trace!(
+                message = "Removing initial BOM bytes from the final output while decoding to utf8.",
+                from_encoding = %self.inner.encoding().name()
+            );
+            output.slice(BOM_UTF8_LEN..)
+        } else {
+            output
+        }
+    }
+}
+
+/// Helps transcoding to the specified encoding from utf8
+pub struct Encoder {
+    buffer: [u8; BUFFER_SIZE],
+    output: BytesMut,
+    inner: encoding_rs::Encoder,
+    // Useful for tracking whether the encoder's encoding is utf-16 (and when it
+    // is, its variety). Since encoding_rs does not have encoders for utf-16,
+    // this is necessary:
+    // https://docs.rs/encoding_rs/0.8.26/encoding_rs/index.html#utf-16le-utf-16be-and-unicode-encoding-schemes
+    utf16_encoding: Option<UTF16Encoding>,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum UTF16Encoding {
+    LE, // little-endian
+    BE, // big-endian
+}
+
+impl Encoder {
+    pub fn new(encoding: &'static Encoding) -> Self {
+        Self {
+            buffer: [0; BUFFER_SIZE],
+            output: BytesMut::new(),
+            inner: encoding.new_encoder(),
+            utf16_encoding: Self::get_utf16_encoding(encoding),
+        }
+    }
+
+    fn get_utf16_encoding(encoding: &'static Encoding) -> Option<UTF16Encoding> {
+        match encoding.name() {
+            "UTF-16LE" => Some(UTF16Encoding::LE),
+            "UTF-16BE" => Some(UTF16Encoding::BE),
+            _ => None,
+        }
+    }
+
+    fn encode_from_utf8_to_utf16(&mut self, input: &str, variant: &UTF16Encoding) -> Bytes {
+        let to_bytes_func = match variant {
+            UTF16Encoding::LE => u16::to_le_bytes,
+            UTF16Encoding::BE => u16::to_be_bytes,
+        };
+
+        for utf16_value in input.encode_utf16() {
+            self.output.extend_from_slice(&to_bytes_func(utf16_value));
+        }
+
+        self.output.split().freeze()
+    }
+
+    pub fn encode_from_utf8(&mut self, input: &str) -> Bytes {
+        // alternate logic if the encoder is for a utf-16 encoding variant
+        if let Some(variant) = self.utf16_encoding {
+            return self.encode_from_utf8_to_utf16(input, &variant);
+        }
+
+        let mut total_read_from_input = 0;
+        let mut total_had_errors = false;
+
+        loop {
```

Comment:
This looks like almost exactly the same loop as in `decode_to_utf8`. Could these be unified into one implementation used twice?

Reply:
I can put the loop logic in a macro and share it in both places, if you like that. I can also try and build something generic to capture the logic, but the effort/indirection may not be worth it since it will be used only in two places (plus the loop logic here is not that long).

Reply:
I would discourage the macro approach. I was thinking a function that would be generic on the encode/extend parameters, either as closures or something else.

You are correct the logic isn't long, though, so it's probably not worth a large effort if it turns into something complex.